nexus_constructor/file_writer_ctrl_window.py

```diff
@@ -0,0 +1,333 @@
+    :param line_edit: The line edit object to apply the validation to.
+    :param is_valid: Whether the line edit field contains valid text
+    :param suggestion_callable: A callable that returns the suggested alternative if not valid.
+    :param tooltip_on_accept: Tooltip to display if line edit is valid.
+    :param tooltip_on_reject: Tooltip to display if line edit is invalid.
+    :return: None.
+    """
+    colour = "white" if is_valid else "red"
+    palette = line_edit.palette()
+    palette.setColor(line_edit.backgroundRole(), QColor(colour))
+    line_edit.setPalette(palette)
+    line_edit.setAutoFillBackground(True)
+    if "Suggestion" in tooltip_on_reject and callable(suggestion_callable):
+        tooltip_on_reject += suggestion_callable()
+    line_edit.setToolTip(tooltip_on_accept) if is_valid else line_edit.setToolTip(
+        tooltip_on_reject
+    )
+
+
+def extract_addr_and_topic(in_string):
+    correct_string_re = re.compile(
+        "(\s*((([^/?#:]+)+)(:(\d+))?)/([a-zA-Z0-9._-]+)\s*)"
+    )  # noqa: W605
+    match_res = re.match(correct_string_re, in_string)
+    if match_res is not None:
+        return match_res.group(2), match_res.group(7)
+    return None
+
+
+class FileWriter:
+    def __init__(self, name, row):
+        self.name = name
+        self.row = row
+        self.last_time = 0
+
+
+class File:
+    def __init__(self, name, row, writer_id, job_id):
+        self.name = name
+        self.row = row
+        self.writer_id = writer_id
+        self.job_id = job_id
+        self.last_time = 0
+
+
+class FileWriterCtrl(Ui_FilewriterCtrl, QMainWindow):
+    def __init__(self, instrument: Instrument):
+        super().__init__()
+        self.settings = QSettings("ess", "nexus-constructor")
+        self.instrument = instrument
+        self.setupUi()
+        self.updateFileNamePlaceholder()
+
+    def setupUi(self):
+        super().setupUi(self)
+        self.statusBrokerLed = Led(self)
+        self.statusTopicLayout.addWidget(self.statusBrokerLed)
+        self.statusBrokerLed.turn_on(False)
+
+        self.statusBrokerEdit.textChanged.connect(self.onTextChanged)
+        self.statusBrokerChangeTimer = QTimer()
+        self.statusBrokerChangeTimer.setSingleShot(True)
+        self.statusBrokerChangeTimer.timeout.connect(self.onTextChangedTimer)
+        self.status_consumer = None
+
+        self.commandBrokerLed = Led(self)
+        self.commandBrokerLayout.addWidget(self.commandBrokerLed)
+        self.commandBrokerLed.turn_on(False)
+        self.commandBrokerEdit.textChanged.connect(self.onCommandBrokerTextChanged)
+        self.commandBrokerChangeTimer = QTimer()
+        self.commandBrokerChangeTimer.setSingleShot(True)
+        self.commandBrokerChangeTimer.timeout.connect(
+            self.onCommandBrokerTextChangeTimer
+        )
+        self.command_producer = None
+
+        self.fileNameLineEdit.textChanged.connect(self.onFileNameChange)
```

Comment:
I am wondering if some of the code in filewriter_command_dialog can be used, possibly even the whole thing? it seems like it could at least be inherited from, seeing as it's doing the same kind of thing?